### PR TITLE
Feature/pipeline wait actionexecutor

### PR DESCRIPTION
**Pull request by chengjoey**
#### What type of this PR

Add one of the following kinds:
/kind feature

#### What this PR does / why we need it:
add wait action executor
make sleep action executing in memory instead of k8sjob

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/all?id=206507&issueFilter__urlQuery=eyJzdGF0ZUJlbG9uZ3MiOlsiT1BFTiIsIldPUktJTkciLCJXT05URklYIiwiUkVPUEVOIl0sImFzc2lnbmVlSURzIjpbIjEwMDEyMDUiXX0%3D&issueTable__urlQuery=eyJwYWdlTm8iOjF9&issueViewGroup__urlQuery=eyJ2YWx1ZSI6ImthbmJhbiIsImNoaWxkcmVuVmFsdWUiOnsia2FuYmFuIjoiZGVhZGxpbmUifX0%3D&iterationID=-1&type=TASK)

#### Specified Reviewers:

/assign @your-reviewer

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |  pipeline add wait actionexecutor            |
| 🇨🇳 中文    | 增加wait actionexecutor             |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by chengjoey** [comment]
image
image

**Comment by sfwn** [comment]
cost_time_sec is not very precise.

File: apistructs/autotest_scene.go [hunk 1]
```diff
@@ -35,7 +35,7 @@ type AutoTestRunStep struct {
 }
 
 type AutoTestRunWait struct {
-	WaitTime int `json:"waitTime"`
+	WaitTime int `json:"waitTime" env:"ACTION_WAIT_TIME"`
```
Comment (sfwn) [inline apistructs/autotest_scene.go:38]: missing unit.

File: modules/pipeline/pipengine/actionexecutor/plugins/wait/wait.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+		return false, false, fmt.Errorf("invalid status when query task exist")
+	}
+}
+
+func (w *Wait) Create(ctx context.Context, task *spec.PipelineTask) (interface{}, error) {
+	return nil, nil
+}
+
+func (w *Wait) Start(ctx context.Context, task *spec.PipelineTask) (interface{}, error) {
+	created, started, err := w.Exist(ctx, task)
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/wait/wait.go:76]: Why need check Exist here?

File: modules/pipeline/pipengine/actionexecutor/plugins/wait/wait.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+	}
+	return envs
+}
+
+func (w *Wait) getWaitSec(task *spec.PipelineTask) (int, error) {
+	envs := mergeEnvs(task)
+
+	var cfg apistructs.AutoTestRunWait
+	if err := envconf.Load(&cfg, envs); err != nil {
+		return 0, errors.Errorf("Failed to get wati time, err: %v", err)
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/wait/wait.go:185]: lowercase at beginning.

Comment (sfwn) [reply modules/pipeline/pipengine/actionexecutor/plugins/wait/wait.go:185]: check other error.

File: modules/pipeline/pipengine/actionexecutor/plugins/wait/wait.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	if executorCh == nil {
+		return nil, errors.Errorf("wait: failed to get exector channel, pipelineID: %d, taskID: %d", task.PipelineID, task.ID)
+	}
+
+	waitSec, err := w.getWaitSec(task)
+	if err != nil {
+		return nil, err
+	}
+
+	timer := time.NewTimer(time.Duration(waitSec) * time.Second)
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/wait/wait.go:105]: Bad.
How about already passed the waitTime if pipeline broken for a long time?

File: modules/pipeline/pipengine/reconciler/listen.go [hunk 5]
```diff
@@ -188,6 +188,8 @@ func makeContextForPipelineReconcile(pipelineID uint64) context.Context {
 	pCtx := context.WithValue(context.Background(), ctxKeyPipelineID, pipelineID)
 	exitCh := make(chan struct{})
 	pCtx = context.WithValue(pCtx, ctxKeyPipelineExitCh, exitCh)
+	executorCh := make(chan interface{})
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/listen.go:191]: executorDoneCh

File: modules/pipeline/pipengine/reconciler/taskrun/framework.go [hunk 6]
```diff
@@ -159,6 +161,14 @@ func (tr *TaskRun) waitOp(itr TaskOp, o *Elem) (result error) {
 		// aop
 		_ = aop.Handle(aop.NewContextForTask(*tr.Task, *tr.P, itr.TuneTriggers().AfterProcessing))
 
+	case data := <-o.ExecutorCh:
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/taskrun/framework.go:164]: use fallthrough before o.DoneCh

File: modules/pipeline/pipengine/reconciler/listen.go [hunk 7]
```diff
@@ -188,6 +188,8 @@ func makeContextForPipelineReconcile(pipelineID uint64) context.Context {
...
+	pCtx = context.WithValue(pCtx, spec.CtxExecutorChKey, executorCh)
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/listen.go:192]: executor is action level, not pipeline level.

**Review by sfwn — CHANGES_REQUESTED** [review]

**Comment by chengjoey** [comment]
> cost_time_sec is not very precise.

cost_time_sec is float to int, missed Precision

File: apistructs/autotest_scene.go (hunk 1 above)
Comment (chengjoey) [reply apistructs/autotest_scene.go:38]: added

File: modules/pipeline/pipengine/actionexecutor/plugins/wait/wait.go (hunk 4 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/wait/wait.go:105]: when execute status interface, will get success status if passed the waitTIme

File: modules/pipeline/pipengine/reconciler/listen.go (hunk 5 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/reconciler/listen.go]: adjusted

File: modules/pipeline/pipengine/reconciler/taskrun/framework.go (hunk 6 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/reconciler/taskrun/framework.go]: select doesn't have fallthrough grammar

File: modules/pipeline/pipengine/reconciler/listen.go (hunk 7 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/reconciler/listen.go]: move the executor chan to traskRun framework

**Review by sfwn — CHANGES_REQUESTED** [review]

**Review by sfwn — DISMISSED** [review]

**Pull request merged**